r.earthworks: major update with quadtree segmentation and new slope functions

baharmon opened the pull request:
This is a major update for `r.earthworks` that adds new features and handles large input elevation rasters more robustly and efficiently. This update includes new features, refactoring, fixing typos, updated docs, etc. Due to changes in input parameters, all documentation, notebooks, tests, and tutorials were updated. 

### Quadtree segmentation
When `r.earthworks` is run with a large input elevation raster and a large number of input coordinates, the computational region is recursively divided in quadrants until a threshold of coordinates per quadrant is met. The computational region for each quadrants is grown by a border parameter to avoid edge effects. The threshold and border are new input parameters in a new quadtree tab of the tool's interface. Segmentation can be disabled with flag `d`.

### New slope functions
* Generalized logistic
* Gaussian
* Cauchy-Lorentz
* Quadratic
* Cubic

### Issues
* A new tutorial covering slope functions is needed. 
* The border for quadtree segmentation is set to a large enough value for common cases, but users may need to adjust the parameter if their earthworks exceed quadrants with default borders. Perhaps the border size could be calculated dynamically based on the max run of the slope per quadrant.
* Many of the slope functions have long tailed distributions that cause tiny elevation changes far away. The output rasters are currently double precision (DCELL) even for a single precision (FCELL) input, so these elevation changes can be very tiny and very far away.

baharmon committed:
r_earthworks: intial commitfor terrain modeling addon

baharmon committed:
r_earthworks: intial commitfor terrain modeling addon

baharmon committed:
r_earthworks: intial release  for terrain modeling addon

baharmon committed:
cleared notebook cells

baharmon committed:
Merge branch 'grass8' into r_earthworks

baharmon committed:
r.earthworks: readme

baharmon committed:
Merge branch 'OSGeo:grass8' into r_earthworks

baharmon committed:
:wqgerge branch 'r_earthworks' of github.com:baharmon/grass-addons into r_earthworks

baharmon committed:
html citation

baharmon committed:
adaptive region for major speed increase

baharmon committed:
updated adaptive region

petrasovaa committed:
resolve conflict

petrasovaa committed:
typos

baharmon committed:
Merge branch 'OSGeo:grass8' into r_earthworks

baharmon committed:
r.earthworks: major update with quadtree segmentation and new slope functions

baharmon committed:
precommit

petrasovaa commented on src/raster/r.earthworks/r.earthworks.html:
```diff
@@ -76,27 +87,35 @@ <h3>Outputs</h3>
 then the earthworking operations may be incomplete
 with artifacts along their edges.
-If artifacts occur, then increase the size of the border.
+If artifacts occurs, then increase the size of the border.
```
occur

petrasovaa reviewed (commented):
I am not able to give you a thorough review right now, so perhaps fix the formatting issues and we can merge it and I can look at it later.

baharmon committed:
update

baharmon commented:
@petrasovaa I forgot to use precommit for the first commit, but ran it and fixed the formatting issues in commit [14ddf04](https://github.com/OSGeo/grass-addons/pull/1458/commits/14ddf0441877e5afd69798404e3050e623c28012). I just corrected the typo in the docs with commit [1f0abf7](https://github.com/OSGeo/grass-addons/pull/1458/commits/1f0abf7d05c4bc5a08055a2efdba400d64f9f99b). So I think it is ready for merging now. A review later would be great.
